rfcs/records.md

```diff
@@ -0,0 +1,201 @@
+
+The difference between records and tables is that when creating a record, the VM allocates space for all values of all keys without having to duplicate the key data;
+since the structure of a record is immutable even if the contents isn't, this allows a much more efficient representation. A 6-field record will take estimated `8+32+6*16=136`
+bytes without implementation heroics, which is more than twice as efficient as table storage. For applications that use many objects this has potential to halve
+the memory footprint.
+
+Record fields use flexible types at runtime (we always allocate space for TValue and don't restrict writes into the record to a given type). In the future, we may
+introduce support for packed records where the table definition must use types and writes that don't abide by these types will trigger an error.
+
+> TODO: It's going to be difficult to migrate to packed records. Is there anything we can do right now to keep this possibility open without enforcing types at runtime?
+
+The rest of this proposal goes into syntax and semantics. The goal of this proposal is to solve the problem of object storage both for simple objects and for classes -- that is, if we add records we won't need to add classes.
+
+### Record type
+
+A record is a collectable object that stores the field values as a inline array of values (TValues) as well as a pointer to the shape. Shape is a table
+that stores various lookup data as implementation details as well as metafields. `type(r)` is `"record"`; `getmetatable(r)` can be used to retrieve the shape.
+
+Being a collectable object, records use raw equality by default when comparing using `==` or hashing; equality behavior can be overridden via `__eq`. From this
+perspective, one could think of records as a user-defined userdata type as opposed to a host-defined userdata type: both typically expose a strict set of fields,
+both are heap-allocated, both use contiguous storage.
+
+Shape contains field lookup data in extra storage that's only allocated for shape tables, as well as the regular table fields. The field lookup data is internal and
+immutable; for example, it might contain a string->index dictionary to be able to quickly locate fields in internal storage.
+
+> TODO: Still not fully set on whether we can get by without a first class shape type.
+
+Reading and writing fields from a record uses `.` or `[]` operator; like tables, passing the field name returns the field value. Unlike tables, if the field is not
+present in the table, the error is raised. This is in constrast with tables where `nil` is returned for unknown keys upon read; records are meant to be stricted than
+tables and as such returning `nil` will mask valuable errors, and make it more difficult to be strict about the types of the result.
+
+Invoking methods with `:` desugars into `getmetatable(obj).__index.method(obj, args)` instead of the usual `obj.method(obj, args)`. This is important because it allows to
```

Comment:
Yeah that's correct. The big difference is that for cases when x is a table, our fast path requires a lot of preconditions; the proposed structure enables a much faster fast path for records.

Reply:
Does it? You still need the slow hash table lookup path in case you predicted M incorrectly, don't you? The code for records if you have predicted metatable M0 is
> if x is a record with metatable M0
> then call f0(x)
> otherwise if x is a record with metatable M
> then lookup m in M, check that the result is a function pointer f, and call f(x)

You still need the slow path for the case where M isn't M0 don't you? You can do clever source-to-source transforms to have specialized versions where we know M is M0 (though hardware branch prediction might be good enough it's not worth bothering with).

All this shape analysis optimization applies just as much to tables (if we optimized them) as it does to records.

Reply:
> Does it?

Yes because the semantics are defined in such a way that methods aren't looked up in the actual table. Please refer to the source of the interpreter, particularly the NAMECALL opcode.

> All this shape analysis optimization applies just as much to tables (if we optimized them) as it does to records.

The critical difference between records and tables doesn't lie in the predictive slots - we already do that for tables. The RFC explicitly spells out the efficiency improvements as being the goal number 3 for this reason. To make table shape inference practical you need to learn how to morph table shapes at runtime. This is more complicated - you need to keep a cache of shapes, track liveness of entries in this cache, and have an invalidation path. Often invalidations will lead to reallocations, so you need to minimize them. As such you need to trace the invalidations back to the origin of table allocations. There are also restrictions on the object placement on the heap, records can be allocated as a single allocation because they never need to be reallocated.